File: gunkims/leetcode/string/3. Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int max = 0;
        int i = 0;
        int j = 0;

        Set<Character> cache = new HashSet<>();

        while (j < s.length()) {
            char ch = s.charAt(j);

            if (cache.contains(ch) == false) {
                cache.add(ch);
                max = Math.max(max, j - i + 1);
                j++;
            } else {
                cache.remove(s.charAt(i));
                i++;
            }
        }
        return max;
    }
}

File: anb/leetcode/string/3. Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int l = 0, r = 0, ans = 0, tmp, N = s.length(), c[256]={};
        while(l < N){
            if(r < N && !c[s[r]]) {
                c[s[r]] = 1;
                r++;
            } else {
                c[s[l]] = 0;
                l++;
            }
            ans = max(ans, r - l);
        }
        return ans;
    }
};